src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -74,7 +74,7 @@ typesOf ::
   [(Ident, Expr)] ->
   m [(Ident, (Expr, Type))]
 typesOf bindingGroupType moduleName vals = do
-  ((tys, subst), warnings) <- reflectErrors . censor (const mempty) . reifyErrors . listen $ liftUnify $ do
+  ((tys, subst), warnings) <- reflectErrors . censor (const mempty) . reifyErrors . listen $ withEmptySubstitution $ do
```

Comment:
It's something to do with types in error messages, but I forget the exact details. Somehow, the substitution was going missing in the presence of an error, so I had to do some manual interleaving of effects. It might be possible to do a better job by breaking up the typechecker state into different parts.